src/UpgradeRule/PHP/Visitor/ClassWarningsVisitor.php

```diff
@@ -0,0 +1,76 @@
+<?php
+
+namespace SilverStripe\Upgrader\UpgradeRule\PHP\Visitor;
+
+use PhpParser\Node;
+use PhpParser\Node\Stmt\Class_;
+use PhpParser\Node\Expr\New_;
+use PhpParser\Node\Expr\StaticCall;
+use SilverStripe\Upgrader\Util\Warning;
+use SilverStripe\Upgrader\Util\ApiChangeWarningSpec;
+
+/**
+ * Relies on {@link SymbolContextVisitor} to annotate the
+ * 'symbolContext' attribute of nodes.
+ *
+ * @package SilverStripe\Upgrader\UpgradeRule
+ */
+class ClassWarningsVisitor extends WarningsVisitor
+{
+
+    public function enterNode(Node $node)
+    {
+        parent::enterNode($node);
+
+        $isClassNode = (
+            $node instanceof Class_ ||
+            $node instanceof StaticCall ||
+            $node instanceof New_
+        );
+        if ($isClassNode) {
+            foreach ($this->specs as $spec) {
+                if (!$this->matchesSpec($node, $spec)) {
+                    continue;
+                }
+
+                $this->addWarning($node, $spec);
+            }
+        }
+    }
+
+    /**
+     * @param Node $node
+     * @param ApiChangeWarningSpec $spec
+     * @return bool
+     */
+    protected function matchesSpec(Node $node, ApiChangeWarningSpec $spec)
+    {
+        $symbol = $spec->getSymbol();
+        $context = $node->getAttribute('symbolContext');
+
+        $class = '';
+
+        if (isset($node->name)) {
+            $class = $node->name;
```

Comment:
This is not immediately obvious, so it's totally not your fault, but `$node->name` MAY contain an object and not a string, but this object will resolve to the correct string with `(string)$node->name`, so you will need to enforce this here.

See `NameResolver::resolveOtherName()` for reference.

Reply:
Will need to be addressed elsewhere too. :)

Reply:
Yeah it's a bit annoying that there's no clear API for this - I'm casting them throughout the code now